File: be/app/schemas/economics.py
import uuid
from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, field_validator


_INCOME_CATEGORIES = {
    "venta_leche", "venta_animal", "venta_cria", "subsidio", "otro_ingreso",
}
_EXPENSE_CATEGORIES = {
    "compra_insumo", "compra_animal", "servicio_veterinario",
    "transporte", "personal", "mantenimiento", "impuesto", "otro_gasto",
}


class EconomicRecordCreate(BaseModel):
    record_type: str
    category: str
    amount: Decimal
    record_date: date
    description: str | None = None
    reference_type: str | None = None
    reference_id: uuid.UUID | None = None

    @field_validator("record_type")
    @classmethod
    def validate_record_type(cls, v: str) -> str:
        allowed = {"ingreso", "egreso"}
        if v.lower() not in allowed:
            raise ValueError(f"Tipo invalido. Opciones: {', '.join(sorted(allowed))}")
        return v.lower()

    @field_validator("category")
    @classmethod
    def validate_category(cls, v: str) -> str:
        v = v.lower()
        all_categories = _INCOME_CATEGORIES | _EXPENSE_CATEGORIES
        if v not in all_categories:
            raise ValueError(f"Categoria invalida. Opciones: {', '.join(sorted(all_categories))}")
        return v

    @field_validator("amount")
    @classmethod
    def validate_amount(cls, v: Decimal) -> Decimal:
        if v <= 0:
            raise ValueError("El monto debe ser mayor a 0")
        return v


class EconomicRecordUpdate(BaseModel):
    category: str | None = None
    amount: Decimal | None = None
    record_date: date | None = None
    description: str | None = None


class EconomicRecordResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    farm_id: uuid.UUID
    record_type: str
    category: str
    amount: Decimal
    record_date: date
    description: str | None
    reference_type: str | None
    reference_id: uuid.UUID | None
    registered_by: uuid.UUID
    created_at: datetime
    updated_at: datetime


class EconomicIndicators(BaseModel):
    total_income: Decimal
    total_expense: Decimal
    balance: Decimal
    income_by_category: dict[str, Decimal]
    expense_by_category: dict[str, Decimal]
    monthly_income: Decimal
    monthly_expense: Decimal
    monthly_balance: Decimal
    avg_income_per_day: Decimal
    avg_expense_per_day: Decimal

File: fe/src/api/economics.ts
import api from "./axios";

export interface EconomicRecordCreate {
  record_type: "ingreso" | "egreso";
  category: string;
  amount: number;
  record_date: string;
  description?: string | null;
  reference_type?: string | null;
  reference_id?: string | null;
}

export interface EconomicRecordUpdate {
  category?: string;
  amount?: number;
  record_date?: string;
  description?: string | null;
}

export interface EconomicRecordResponse extends EconomicRecordCreate {
  id: string;
  farm_id: string;
  registered_by: string;
  created_at: string;
  updated_at: string;
}

export interface EconomicIndicators {
  total_income: number;
  total_expense: number;
  balance: number;
  income_by_category: Record<string, number>;
  expense_by_category: Record<string, number>;
  monthly_income: number;
  monthly_expense: number;
  monthly_balance: number;
  avg_income_per_day: number;
  avg_expense_per_day: number;
}

const base = (farmId: string) => `/farms/${farmId}/economics`;

export async function createRecord(
  farmId: string,
  data: EconomicRecordCreate,
): Promise<EconomicRecordResponse> {
  const response = await api.post<EconomicRecordResponse>(base(farmId), data);
  return response.data;
}

export async function listRecords(
  farmId: string,
  params?: {
    record_type?: string;
    category?: string;
    date_from?: string;
    date_to?: string;
  },
): Promise<EconomicRecordResponse[]> {
  const response = await api.get<EconomicRecordResponse[]>(base(farmId), {
    params,
  });
  return response.data;
}

export async function getRecord(
  farmId: string,
  recordId: string,
): Promise<EconomicRecordResponse> {
  const response = await api.get<EconomicRecordResponse>(
    `${base(farmId)}/${recordId}`,
  );
  return response.data;
}

export async function updateRecord(
  farmId: string,
  recordId: string,
  data: EconomicRecordUpdate,
): Promise<EconomicRecordResponse> {
  const response = await api.put<EconomicRecordResponse>(
    `${base(farmId)}/${recordId}`,
    data,
  );
  return response.data;
}

export async function deleteRecord(
  farmId: string,
  recordId: string,
): Promise<void> {
  await api.delete(`${base(farmId)}/${recordId}`);
}

export async function getIndicators(
  farmId: string,
  params?: { date_from?: string; date_to?: string },
): Promise<EconomicIndicators> {
  const response = await api.get<EconomicIndicators>(
    `${base(farmId)}/indicators`,
    { params },
  );
  return response.data;
}
